src/sources/datadog/agent/mod.rs

```diff
@@ -159,13 +161,38 @@ impl SourceConfig for DatadogAgentConfig {
     }
 
     fn outputs(&self) -> Vec<Output> {
+        let definition = match self.decoding {
+            // See: `LogMsg` struct.
+            DeserializerConfig::Bytes => schema::Definition::empty()
+                .required_field("message", Kind::bytes(), Some("message"))
+                .required_field("status", Kind::bytes(), Some("severity"))
+                .required_field("timestamp", Kind::integer(), Some("timestamp"))
+                .required_field("hostname", Kind::bytes(), Some("host"))
+                .required_field("service", Kind::bytes(), None)
+                .required_field("ddsource", Kind::bytes(), None)
+                .required_field("ddtags", Kind::bytes(), None)
+                .merge(self.decoding.schema_definition()),
+
+            // JSON deserializer can overwrite existing fields at runtime, so we have to treat
+            // those events as if there is no known type details we can provide, other than the
+            // details provided by the generic JSON schema definition.
+            DeserializerConfig::Json => self.decoding.schema_definition(),
+
+            // Syslog deserializer allows for arbritrary "structured data" that can overwrite
+            // existing fields, similar to the JSON deserializer.
+            //
+            // See also: https://datatracker.ietf.org/doc/html/rfc5424#section-6.3
+            #[cfg(feature = "sources-syslog")]
+            DeserializerConfig::Syslog => self.decoding.schema_definition(),
+        };
+
         if self.multiple_outputs {
             vec![
                 Output::from((METRICS, DataType::Metric)),
-                Output::from((LOGS, DataType::Log)),
+                Output::from((LOGS, DataType::Log)).with_schema_definition(definition),
             ]
         } else {
-            vec![Output::default(DataType::Any)]
+            vec![Output::default(DataType::Any).with_schema_definition(definition)]
```

Comment:
This is a _little bit_ awkward right now, as it means we store the schema information separately from `DataType`, which means you can end up with a `DataType::Metric` that has schema information embedded, even though we don't support schemas for metrics yet (but will in the near future).

The churn caused by updating `DataType` was too much for now, however, so I opted for this _slightly_ leaky API. We'll revisit this once we start implementing schema information for more sources, and it has become more obvious what the right approach is here.